File: DecompiledGameScripts/cyberpunk/devices/utilities/weaponTraining/weaponTraining.swift

public class WeaponTraining extends InteractiveDevice {

  @runtimeProperty("customEditor", "TweakDBGroupInheritance;RPGActionRewards.ProficiencyReward")
  protected edit let m_rewardRecord: TweakDBID;

  protected edit const let m_weaponTypes: [gamedataItemType];

  @default(WeaponTraining, 30)
  protected edit let m_limitOfHits: Int32;

  protected let m_amountOfHits: Int32;

  protected cb func OnRequestComponents(ri: EntityRequestComponentsInterface) -> Bool {
    super.OnRequestComponents(ri);
  }

  protected cb func OnTakeControl(ri: EntityResolveComponentsInterface) -> Bool {
    super.OnTakeControl(ri);
    this.m_controller = EntityResolveComponentsInterface.GetComponent(ri, n"controller") as WeaponTrainingController;
  }

  protected cb func OnHitEvent(hit: ref<gameHitEvent>) -> Bool {
    let weaponRecord: wref<Item_Record>;
    let attackData: ref<AttackData> = hit.attackData;
    if attackData.GetInstigator().IsPlayer() && !hit.target.IsPlayer() {
      weaponRecord = TweakDBInterface.GetItemRecord(ItemID.GetTDBID(attackData.GetWeapon().GetItemID()));
      if this.m_amountOfHits < this.m_limitOfHits && this.MatchWeaponItemType(weaponRecord.ItemType().Type()) && TDBID.IsValid(this.m_rewardRecord) {
        this.AwardRewardXP(attackData.GetInstigator(), hit.target.GetEntityID());
        this.m_amountOfHits += 1;
      };
    };
  }

  protected final func MatchWeaponItemType(type: gamedataItemType) -> Bool {
    let i: Int32 = 0;
    while i < ArraySize(this.m_weaponTypes) {
      if Equals(this.m_weaponTypes[i], type) {
        return true;
      };
      i += 1;
    };
    return false;
  }

  protected final func AwardRewardXP(instigator: wref<GameObject>, target: EntityID) -> Void {
    let expArr: array<wref<XPPoints_Record>>;
    let expType: gamedataProficiencyType;
    let expValue: Float;
    let hitModifier: Float;
    let i: Int32;
    let quantityMods: array<wref<StatModifier_Record>>;
    let rewardRecord: ref<RewardBase_Record> = TweakDBInterface.GetRewardBaseRecord(this.m_rewardRecord);
    rewardRecord.Experience(expArr);
    hitModifier = 1.00 - Cast<Float>(this.m_amountOfHits) / (Cast<Float>(this.m_limitOfHits) + 1.00);
    i = 0;
    while i < ArraySize(expArr) {
      ArrayClear(quantityMods);
      expArr[i].QuantityModifiers(quantityMods);
      expValue = RPGManager.CalculateStatModifiers(quantityMods, this.GetGame(), instigator, Cast<StatsObjectID>(target));
      expType = expArr[i].Type().Type();
      RPGManager.AwardXP(this.GetGame(), expValue * hitModifier, expType);
      i += 1;
    };
  }

  protected const func HasAnyDirectInteractionActive() -> Bool {
    return true;
  }

  public const func DeterminGameplayRole() -> EGameplayRole {
    return EGameplayRole.GenericRole;
  }
}

public class WeaponTrainingController extends ScriptableDeviceComponent {

  public const func GetPS() -> ref<GameComponentPS> {
    return this.GetBasePS();
  }
}

File: Definitions/Game_Definitions/classes/WeaponTraining.lua
---@meta
---@diagnostic disable

---@class WeaponTraining : InteractiveDevice
---@field rewardRecord TweakDBID
---@field weaponTypes gamedataItemType[]
---@field limitOfHits Int32
---@field amountOfHits Int32
WeaponTraining = {}

---@return WeaponTraining
function WeaponTraining.new() return end

---@param props table
---@return WeaponTraining
function WeaponTraining.new(props) return end

---@param hit gameeventsHitEvent
---@return Bool
function WeaponTraining:OnHitEvent(hit) return end

---@param ri entEntityRequestComponentsInterface
---@return Bool
function WeaponTraining:OnRequestComponents(ri) return end

---@param ri entEntityResolveComponentsInterface
---@return Bool
function WeaponTraining:OnTakeControl(ri) return end

---@param instigator gameObject
---@param target entEntityID
function WeaponTraining:AwardRewardXP(instigator, target) return end

---@return EGameplayRole
function WeaponTraining:DeterminGameplayRole() return end

---@return Bool
function WeaponTraining:HasAnyDirectInteractionActive() return end

---@param type gamedataItemType
---@return Bool
function WeaponTraining:MatchWeaponItemType(type) return end

